pilot/pkg/networking/core/cluster_builder.go

```diff
@@ -452,39 +452,35 @@ func (cb *ClusterBuilder) buildInboundCluster(clusterPort int, bind string,
 	return localCluster
 }
 
-// buildInboundPassthroughClusters builds passthrough clusters for inbound.
-func (cb *ClusterBuilder) buildInboundPassthroughClusters() []*cluster.Cluster {
-	// ipv4 and ipv6 feature detection. Envoy cannot ignore a config where the ip version is not supported
-	clusters := make([]*cluster.Cluster, 0, 2)
-	if cb.supportsIPv4 {
-		inboundPassthroughClusterIpv4 := cb.buildDefaultPassthroughCluster()
-		inboundPassthroughClusterIpv4.Name = util.InboundPassthroughClusterIpv4
-		inboundPassthroughClusterIpv4.Filters = nil
-		inboundPassthroughClusterIpv4.UpstreamBindConfig = &core.BindConfig{
-			SourceAddress: &core.SocketAddress{
-				Address: InboundPassthroughBindIpv4,
-				PortSpecifier: &core.SocketAddress_PortValue{
-					PortValue: uint32(0),
-				},
-			},
-		}
-		clusters = append(clusters, inboundPassthroughClusterIpv4)
-	}
-	if cb.supportsIPv6 {
-		inboundPassthroughClusterIpv6 := cb.buildDefaultPassthroughCluster()
-		inboundPassthroughClusterIpv6.Name = util.InboundPassthroughClusterIpv6
-		inboundPassthroughClusterIpv6.Filters = nil
-		inboundPassthroughClusterIpv6.UpstreamBindConfig = &core.BindConfig{
-			SourceAddress: &core.SocketAddress{
-				Address: InboundPassthroughBindIpv6,
-				PortSpecifier: &core.SocketAddress_PortValue{
-					PortValue: uint32(0),
-				},
+// buildInboundPassthroughCluster builds passthrough cluster for inbound.
+func (cb *ClusterBuilder) buildInboundPassthroughCluster() *cluster.Cluster {
+	// We need to set a local bind address, which we will match in iptables to avoid looping back to ourselves.
+	// This needs a per-IP-version, since we cannot bind to IPv4 and send to IPv6 (or the inverse).
+	// Fortunately, Envoy can natively handle this by giving it a local v4 and v6 address, and it will pick which to use for us.
+	src := InboundPassthroughBindIpv4
+	if !cb.supportsIPv4 {
+		src = InboundPassthroughBindIpv6
+	}
+	var extraSrc []*core.ExtraSourceAddress
+	if cb.supportsIPv4 && cb.supportsIPv6 {
+		extraSrc = append(extraSrc, &core.ExtraSourceAddress{
+			Address: &core.SocketAddress{
```

Comment:
It maybe more cleae if using `switch case`

Reply:
we need to do this if only both are true though, not sure how to do that? In rust could do with matching 2 values but not here

Reply:
maybe we pass the ip mode directly here, deprecate these two fields {cb.supportsIPv4, cb.supportsIPv6}

Reply:
will look into it. Might be tricky since this codepath is not covered by the dual stack feature flag (as this has been around since istio 1.3!) which may make it harder to use the IP mode. Not sure, will try it out tomorrow

Reply:
Yeah the IP mode is guarded by the feature flag so would be a bit tricky right now. Once dual stack is stable I think we can do a lot of cleanup

Reply:
make sense